torch_xla/csrc/status.h

```diff
@@ -0,0 +1,81 @@
+#ifndef XLA_TORCH_XLA_CSRC_STATUS_H_
+#define XLA_TORCH_XLA_CSRC_STATUS_H_
+
+#include "absl/status/statusor.h"
+
+namespace torch_xla {
+
+// Creates a new Status instance, appending the current location (e.g. file and
+// line information) to the status message.
+//
+// This should be used whenever we are returning new error status, instead of
+// propagating. Then, if `kEnvShowCppErrorContext` environment variable is set,
+// the location information will be shown.
+#define XLA_ERROR_WITH_LOCATION(STATUS) \
+  ::torch_xla::MaybeWithLocation(std::move(STATUS), __FILE__, __LINE__)
+
+#define _XLA_RETURN_IF_ERROR(EXPR, THEN, ...)                         \
+  do {                                                                \
+    auto _status_var = (EXPR);                                        \
+    if (!_status_var.ok()) {                                          \
+      return ::torch_xla::MaybeWithNewMessage(                        \
+          std::move(_status_var), __FILE__, __LINE__, ##__VA_ARGS__); \
+    }                                                                 \
+    THEN;                                                             \
+  } while (0)
+
+// Propagates `REXPR`, in case it's a non-ok status.
+#define XLA_RETURN_IF_ERROR(REXPR, ...) \
+  _XLA_RETURN_IF_ERROR(REXPR, {}, ##__VA_ARGS__)
+
+// Propagates `REXPR`, in case it's a non-ok status. Otherwise, assign
+// its result to `LHS`.
+#define XLA_ASSIGN_OR_RETURN(LHS, REXPR, ...) \
+  _XLA_RETURN_IF_ERROR(                       \
+      REXPR, { LHS = std::move(_status_var.value()); }, ##__VA_ARGS__)
+
+// Maybe shows location information in the status message.
+//
+// This function assumes that `status` is a non-ok status.
+absl::Status MaybeWithLocation(absl::Status&& status, const char* file,
+                               const int32_t line);
+
+// Maybe replace the current `status` message with `new_message`.
+//
+// This function assumes that `status` is a non-ok status.
+//
+// Rationale: if given, `new_message` has more context, which makes it possible
+// for better error messages to the user.
+//
+// This function also appends file location information to the error message, if
+// `kEnvShowCppErrorContext` is set.
+absl::Status MaybeWithNewMessage(absl::Status&& status, const char* file,
+                                 const int32_t line,
+                                 const std::string_view new_message = "");
+
+template <class T>
+absl::Status MaybeWithNewMessage(absl::StatusOr<T>&& status, const char* file,
```

Comment:
It's surprising that this returns a Status instead of StatusOr. I'd remove it and let the user call `MaybeWithNewMessage(status.status(), ...)`, which is much clearer.

Reply:
What do you think of simply returning `absl::Status`, since it's just a `uintptr_t`?

Reply:
It's not a `uintptr_t`. `Status` is not trivial enough to pass by value. We should avoid copying it.